skimage/feature/register_translation.py

```diff
@@ -58,20 +58,50 @@ def _upsampled_dft(data, upsampled_region_size,
             raise ValueError("number of axis offsets must be equal to input "
                              "data's number of dimensions.")
 
-    col_kernel = np.exp(
-        (-1j * 2 * np.pi / (data.shape[1] * upsample_factor)) *
-        (np.fft.ifftshift(np.arange(data.shape[1]))[:, None] -
-         np.floor(data.shape[1] / 2)).dot(
-             np.arange(upsampled_region_size[1])[None, :] - axis_offsets[1])
-    )
-    row_kernel = np.exp(
-        (-1j * 2 * np.pi / (data.shape[0] * upsample_factor)) *
-        (np.arange(upsampled_region_size[0])[:, None] - axis_offsets[0]).dot(
-            np.fft.ifftshift(np.arange(data.shape[0]))[None, :] -
-            np.floor(data.shape[0] / 2))
-    )
+    # For the 2D case, it is more efficient to use the well-optimized
+    # dot product instead of einsum
```

Comment:
@mattip Could you double check this optimization for us?  I.e., is this the correct approach, giving optimizations that have recently been made inside of NumPy.

Reply:
Benchmarking will give the most authoritative answer, and also will allow you to see if something changes in the future. 

In general, einsum has more parsing overhead, and depends on hand-written SSE routines to vectorize operations, where dot use BLAS-type calls, which could be OpenBLAS, Atlas, or MKL depending on the linalg backend.